### ISSUE TITLE
Race condition in gossip and failure detector

### ISSUE THREAD

**Issue by geobeau** [scylladb/scylla#10547]
Bug was noticed in 4.6.3.

# Problem

After using` nodetool disablegossip` then `nodetool enablegossip` on several nodes of a cluster of 5 nodes, the nodes were unable to see some other nodes as UN again. The only mitigation has been to roll restart the cluster.

After observing the logs, we noticed that a node that had its gossip restarted was only sending echo messages to one node.

Here is an excerpt of the gossip logs:
```
DEBUG 2022-05-11 09:49:37,704 [shard 1] gossip - failure_detector_loop: Send echo to node 10.236.233.16, status = ok
DEBUG 2022-05-11 09:49:37,704 [shard 2] gossip - failure_detector_loop: Send echo to node 10.236.235.16, status = ok
WARN  2022-05-11 09:49:39,064 [shard 0] storage_service - Stopping gossip by operator request
DEBUG 2022-05-11 09:49:39,258 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
DEBUG 2022-05-11 09:49:39,258 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = ok
DEBUG 2022-05-11 09:49:39,704 [shard 3] gossip - failure_detector_loop: Send echo to node 10.236.178.12, status = started
DEBUG 2022-05-11 09:49:39,704 [shard 1] gossip - failure_detector_loop: Send echo to node 10.236.233.16, status = started
DEBUG 2022-05-11 09:49:39,704 [shard 2] gossip - failure_detector_loop: Send echo to node 10.236.235.16, status = started
DEBUG 2022-05-11 09:49:39,719 [shard 3] gossip - failure_detector_loop: Send echo to node 10.236.178.12, status = ok
DEBUG 2022-05-11 09:49:39,720 [shard 1] gossip - failure_detector_loop: Send echo to node 10.236.233.16, status = ok
DEBUG 2022-05-11 09:49:39,720 [shard 2] gossip - failure_detector_loop: Send echo to node 10.236.235.16, status = ok
DEBUG 2022-05-11 09:49:41,721 [shard 0] gossip - failure_detector_loop: previous_live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], current_live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], nodes_down=[], version_before=5, version_after=5
INFO  2022-05-11 09:49:41,721 [shard 0] gossip - failure_detector_loop: Finished main loop
WARN  2022-05-11 09:49:50,349 [shard 0] storage_service - Starting gossip by operator request
INFO  2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Started main loop
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Started new round for node=10.189.22.18 on shard=0, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Started new round for node=10.236.178.12 on shard=1, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Started new round for node=10.236.233.16 on shard=2, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Started new round for node=10.236.235.16 on shard=3, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
DEBUG 2022-05-11 09:49:50,494 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = ok
DEBUG 2022-05-11 09:49:52,494 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
DEBUG 2022-05-11 09:49:52,494 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = ok
DEBUG 2022-05-11 09:49:54,494 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
DEBUG 2022-05-11 09:49:54,495 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = ok
DEBUG 2022-05-11 09:49:56,495 [shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
```
The interesting part is that after the restart, we can see that we send the echo message is only sent to one node despite scylladb scheduling it for the 4 other nodes of the cluster. The other interesting part is that shard 0 is the only one working.

## The gossiper

From shard_0, the gossiper is starting a `failure_detector_loop_for_node` for each peer and assign it to shards in round robin.

```c++
co_await parallel_for_each(boost::irange(size_t(0), nodes.size()), [this, generation_number, live_endpoints_version, &nodes] (size_t idx) {
    const auto& node = nodes[idx];
    auto shard = idx % smp::count;
    logger.debug("failure_detector_loop: Started new round for node={} on shard={}, live_nodes={}, live_endpoints_version={}",
            node, shard, nodes, live_endpoints_version);
    return container().invoke_on(shard, [node, generation_number, live_endpoints_version] (gms::gossiper& g) {
        return g.failure_detector_loop_for_node(node, generation_number, live_endpoints_version);
    });
});
```

In  `failure_detector_loop_for_node`, we can see that the only it can execute and not output a debug log is if `is_enabled()` is not true. However, it is properly set to true in the `start_gossiping` and `do_stop_gossiping functions`.
```c++
future<> gossiper::failure_detector_loop_for_node(gms::inet_address node, int64_t gossip_generation, uint64_t live_endpoints_version) {
    auto last = gossiper::clk::now();
    auto diff = gossiper::clk::duration(0);
    auto echo_interval = std::chrono::milliseconds(2000);
    auto max_duration = echo_interval + std::chrono::milliseconds(_failure_detector_timeout_ms());
    while (is_enabled()) {
        bool failed = false;
        try {
            logger.debug("failure_detector_loop: Send echo to node {}, status = started", node);
            co_await _messaging.send_gossip_echo(netw::msg_addr(node), gossip_generation, max_duration);
            logger.debug("failure_detector_loop: Send echo to node {}, status = ok", node);
        } catch (...) {
            failed = true;
            logger.warn("failure_detector_loop: Send echo to node {}, status = failed: {}", node, std::current_exception());
        }
...
```

## The race condition

Adding debug log makes the race condition quite obvious:
```
[shard 0] gossip - failure_detector_loop: Started new round for node=10.189.22.18 on shard=0, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
[shard 0] gossip - failure_detector_loop: if enabled (true) will send echo to node 10.189.22.18, status = started
[shard 0] gossip - failure_detector_loop: Send echo to node 10.189.22.18, status = started
[shard 0] gossip - failure_detector_loop: Started new round for node=10.236.178.12 on shard=1, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
[shard 0] gossip - failure_detector_loop: Started new round for node=10.236.233.16 on shard=2, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
[shard 0] gossip - failure_detector_loop: Started new round for node=10.236.235.16 on shard=3, live_nodes=[10.189.22.18, 10.236.178.12, 10.236.233.16, 10.236.235.16], live_endpoints_version=5
[shard 2] gossip - failure_detector_loop: if enabled (false) will send echo to node 10.236.233.16, status = started
[shard 3] gossip - failure_detector_loop: if enabled (false) will send echo to node 10.236.235.16, status = started
[shard 1] gossip - failure_detector_loop: if enabled (false) will send echo to node 10.236.178.12, status = started
[shard 2] gossip - failure_detector_loop: gossip is enabled
[shard 4] gossip - failure_detector_loop: gossip is enabled
[shard 5] gossip - failure_detector_loop: gossip is enabled
[shard 7] gossip - failure_detector_loop: gossip is enabled
[shard 11] gossip - failure_detector_loop: gossip is enabled
[shard 9] gossip - failure_detector_loop: gossip is enabled
[shard 13] gossip - failure_detector_loop: gossip is enabled
[shard 3] gossip - failure_detector_loop: gossip is enabled
[shard 6] gossip - failure_detector_loop: gossip is enabled
[shard 10] gossip - failure_detector_loop: gossip is enabled
[shard 8] gossip - failure_detector_loop: gossip is enabled
[shard 1] gossip - failure_detector_loop: gossip is enabled
[shard 12] gossip - failure_detector_loop: gossip is enabled
```

The gossiper is enabled after the `failure_detector_loop` has started to dispatch the `failure_detector_loop_for_node` to other shards. This mean that `failure_detector_loop_for_node` will instantly terminate.

The issue is in gossiper::start_gossiping
```c++
return container().invoke_on_all([] (gms::gossiper& g) {
    logger.debug("failure_detector_loop: gossip is enabled");
    g._enabled = true;
    g._failure_detector_loop_done = g.failure_detector_loop();
});
```

`g.failure_detector_loop();` can be started on shard 0 before have executed the future

# Proposed fix

I opened a PR with a proposed patch but I am scylla/cpp noob

#10548

# Reproducing

Start a cluster with at least 3 nodes. Once all nodes have joined and are UN, disable then enable back the gossip. One node should stay DN despite working well.

**Comment by geobeau** [comment]
Another problem that arises after fixing the failure detector loop is that the `_background_msg` is closed on shutdown but never "opened" again. I added a second commit to make sure it is recreated at startup

**Comment by geobeau** [comment]
Also, I have a patch compatible with 4.6.3 here if you want to backport: https://github.com/criteo-forks/scylla/pull/8/files

**Comment by slivne** [comment]
@kbr-  & @asias the resident gossip experts can you please comment

**Comment by asias** [comment]
@geobeau Thanks for reporting and looking into this. I can reproduce myself. I added a comment in your PR.

**Comment by asias** [comment]
Solved by https://github.com/scylladb/scylla/pull/10548.

**Comment by mykaul** [comment]
@DoronArazii - another case where the bot did not do its job and added 'Backport candidate' label.
We'll need it backported to 5.0.x, I reckon!

**Comment by DoronArazii** [comment]
> @DoronArazii - another case where the bot did not do its job and added 'Backport candidate' label. We'll need it backported to 5.0.x, I reckon!

The commit isn't reference this as "Fixes: #10547".
It's not the bot fault this time :)

**Comment by denesb** [comment]
Already in all live versions, removing label.
